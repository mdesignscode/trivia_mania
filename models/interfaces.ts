export interface IQuestion {
  category: string;
  answers: Array<string>;
  correctAnswer: string;
  id: string;
  question: string;
  difficulty: string;
}
