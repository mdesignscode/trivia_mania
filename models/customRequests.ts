import { IUserStats } from "./interfaces";

export type TDifficulties = "easy" | "medium" | "hard" | ""

export interface IPlayRequest {
  difficulty?: TDifficulties;
  categories?: string[];
  userId?: string;
}

export class PlayRequest extends Request {
  constructor(url: string, options?: RequestInit, customData?: IPlayRequest) {
    super(url, mergeOptions(customData, options));
  }
}

export type TStatsRequest =
  | {
    recordType: 'difficulties';
    userId?: string;
  }
  | {
    recordType: 'categories';
    difficulty: TDifficulties;
    userId?: string;
  };

export class StatsRequest extends Request {
  constructor(url: string, options?: RequestInit, customData?: TStatsRequest) {
    super(url, mergeOptions(customData, options));
  }
}

export interface IGetUserRequest {
  id: string
}

export class GetUserRequest extends Request {
  constructor(url: string, options?: RequestInit, customData?: IGetUserRequest) {
    super(url, mergeOptions(customData, options));
  }
}

export interface IUpdateUserStatsRequest {
  id: string;
  answeredQuestions: string[];
  stats: IUserStats
}

export class UpdateUserStatsRequest extends Request {
  constructor(url: string, options?: RequestInit, customData?: IUpdateUserStatsRequest) {
    super(url, mergeOptions(customData, options));
  }
}

function mergeOptions<RequestType>(customData: RequestType, options?: RequestInit) {
  // Serialize your custom data and add it to the request body
  const customBody = JSON.stringify(customData);

  // Merge the custom options with the provided options
  return {
    ...options,
    method: options?.method || 'POST',
    body: customBody,
  };
}
