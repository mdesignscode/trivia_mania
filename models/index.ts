#!/usr/bin/node

import FileStorage from "./storage/fileStorage";

// initializes the storage object

let storage: FileStorage;
const storageType = process.env.TRIVIA_STORAGE_TYPE;

// if (storageType === 'db') {
//   (async () => {
//     const { DBStorage } = require("./engine/dbStorage");
//     storage = new DBStorage();
//     await storage.reload();
//   })();
// } else {
//   const FileStorage = require("./engine/fileStorage");
//   storage = new FileStorage();
//   storage.reload();
// }

storage = new FileStorage();
storage.reload();

export default storage;
