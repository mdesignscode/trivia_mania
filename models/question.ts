#!/usr/bin/node

import { IQuestion } from "./interfaces";

class Question {
  // Properties
  category: string;
  answers: Array<string>;
  correctAnswer: string;
  id: string;
  question: string;
  difficulty: string;

  // Constructor to initialize properties
  constructor(data: IQuestion) {
    this.category = data.category;
    this.answers = data.answers;
    this.correctAnswer = data.correctAnswer;
    this.id = data.id;
    this.question = data.question;
    this.difficulty = data.difficulty;

    this.shuffleAnswers()
  }

  // Method to check if a given answer is correct
  isCorrectAnswer(answer: string): boolean {
    return answer === this.correctAnswer;
  }

  // Method to shuffle the answer choices (randomize)
  shuffleAnswers(): void {
    for (let i = this.answers.length - 1; i > 0; i--) {
      const index = Math.floor(Math.random() * (i + 1));
      const j = index === i ? i + 1 : index;
      [this.answers[i], this.answers[j]] = [this.answers[j], this.answers[i]];
    }
  }

  // Method to display the question and answer choices
  displayQuestion(): void {
    console.log(`Category: ${this.category}`);
    console.log(`Question: ${this.question}`);
    console.log(`Difficulty: ${this.difficulty}`);
    console.log(`Answers:`);
    this.answers.forEach((answer, index) => {
      console.log(`${index + 1}. ${answer}`);
    });
  }
}

export default Question;
