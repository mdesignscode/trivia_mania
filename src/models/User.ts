import { DataTypes } from "sequelize";
import sequelize from "../utils/sequelize";

const User = sequelize.define(
        'User',
        {
                id: {
                        type: DataTypes.UUID,
                        defaultValue: DataTypes.UUIDV4,
                        primaryKey: true,
                },
                isVerified: {
                        type: DataTypes.BOOLEAN,
                        defaultValue: false,
                },
                password: {
                        type: DataTypes.STRING,
                        allowNull: false,
                },
                username: {
                        type: DataTypes.STRING,
                        allowNull: false,
                },
                email: {
                        type: DataTypes.STRING,
                        allowNull: false,
                },
                avatar: {
                        type: DataTypes.STRING,
                        allowNull: false,
                },
                answeredQuestions: {
                        type: DataTypes.JSON,
                },
                correctAnswered: {
                        type: DataTypes.INTEGER,
                        allowNull: false,
                },
        },
);

export default User;

