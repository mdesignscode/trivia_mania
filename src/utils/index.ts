import { browser } from '$app/environment';

export function decodeHTMLEntities(text: string) {
        if (browser) {
                const element = document.createElement("div");
                element.innerHTML = text;
                return element.textContent || "";
        }
}

export const colorMap: { [key: string]: string } = {
        easy: "green",
        medium: "gold",
        hard: "red",
};

export function capitalize(str: string): string {
	if (!str) return '';
	return str.charAt(0).toUpperCase() + str.slice(1).toLowerCase();
}

