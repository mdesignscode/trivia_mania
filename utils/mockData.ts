import { CategoryStat, DifficultyStat, IUserStats } from "@/models/interfaces";
import { IQuestion } from "@/models/interfaces";

type TStat = Record<string, number>

export const mockInitialProgress: IUserStats = {
  total: {
    correctAnswered: 1,
    answered: 1,
  },
  easy: {
    answered: 1,
    correctAnswered: 1,
  } as DifficultyStat,
  Science: {
    easy: {
      answered: 1,
      correctAnswered: 1,
    } as DifficultyStat,
  } as CategoryStat,
};

// create mock questions
export const mockQuestions: Record<string, IQuestion> = {
  "63805aab-b589-4256-9493-769eecc23966": {
    "category": "Entertainment: Film",
    "answers": [
      "Batman v Superman: Superapocalypse",
      "Batman v Superman: Black of Knight",
      "Batman v Superman: Dawn of Justice",
      "Batman v Superman: Knightfall"
    ],
    "correctAnswer": "Batman v Superman: Dawn of Justice",
    "id": "63805aab-b589-4256-9493-769eecc23966",
    "question": "Which movie released in 2016 features Superman and Batman fighting?",
    "difficulty": "medium"
  },
  "36240604-80aa-4ddb-bb87-5ce345bfd379": {
    "category": "Entertainment: Film",
    "answers": ["R2-D2", "AA-A", "BB-8", "BB-3"],
    "correctAnswer": "BB-8",
    "id": "36240604-80aa-4ddb-bb87-5ce345bfd379",
    "question": "What is the orange and white bot&#039;s name in &quot;Star Wars: The Force Awakens&quot;?",
    "difficulty": "easy"
  },
  "12c4b40f-7fa9-405d-81de-b64c54a0caea": {
    "category": "Entertainment: Music",
    "answers": [
      "James Blunt - &quot;1973&quot;",
      "Prince - &quot;1999&quot;",
      "The Smashing Pumpkins - &quot;1979&quot;",
      "David Bowie - &quot;1984&quot;"
    ],
    "correctAnswer": "The Smashing Pumpkins - &quot;1979&quot;",
    "id": "12c4b40f-7fa9-405d-81de-b64c54a0caea",
    "question": "Which of these songs was released in 1996?",
    "difficulty": "hard"
  },
  "b2d0c4b4-ac68-491c-a561-60848ea34fb6": {
    "category": "Entertainment: Music",
    "answers": [
      "Gimmie! Gimmie! Gimme! (A Man After Midnight)",
      "Night Fever",
      "The Chain",
      "Staying Alive"
    ],
    "correctAnswer": "Gimmie! Gimmie! Gimme! (A Man After Midnight)",
    "id": "b2d0c4b4-ac68-491c-a561-60848ea34fb6",
    "question": "Madonna&#039;s song &quot;Hung Up&quot; includes a piece from which popular 70s song?",
    "difficulty": "medium"
  },
  "83cea789-1c6d-4757-a571-c79a14a9133f": {
    "category": "General Knowledge",
    "answers": ["Liver", "Skin", "Heart", "large Intestine"],
    "correctAnswer": "Skin",
    "id": "83cea789-1c6d-4757-a571-c79a14a9133f",
    "question": "What is the largest organ of the human body?",
    "difficulty": "easy"
  },
  "c051d919-2618-4f4d-bb42-5e13bdf58423": {
    "category": "General Knowledge",
    "answers": ["Vanilla", "Saffron", "Cinnamon", "Cardamom"],
    "correctAnswer": "Saffron",
    "id": "c051d919-2618-4f4d-bb42-5e13bdf58423",
    "question": "What is the world&#039;s most expensive spice by weight?",
    "difficulty": "medium"
  },
  "9289f1e9-f0c7-4fe4-a4f9-755745eb0bc8": {
    "category": "Entertainment: Video Games",
    "answers": ["Diamond/Pearl", "Ruby/Sapphire", "X/Y", "Black/White"],
    "correctAnswer": "Diamond/Pearl",
    "id": "9289f1e9-f0c7-4fe4-a4f9-755745eb0bc8",
    "question": "Which Pokemon generation did the fan-named &quot;Masuda Method&quot; first appear in? ",
    "difficulty": "easy"
  },
  "4db885c9-d4c1-4515-9e20-db71d6557cd6": {
    "category": "Entertainment: Video Games",
    "answers": [
      "Tateyuki Shigaraki",
      "Manfred Von Karma",
      "Gregory Edgeworth",
      "Bansai Ichiyanagi"
    ],
    "correctAnswer": "Bansai Ichiyanagi",
    "id": "4db885c9-d4c1-4515-9e20-db71d6557cd6",
    "question": "In the &quot;Ace Attorney&quot; series, who was the truly responsible of the forging of the autopsy report of the pivotal IS-7 incident?",
    "difficulty": "hard"
  },
}

// mock questions stats
export const mockDifficultyStats: TStat = {
  medium: 3,
  hard: 2,
  easy: 3,
  "all difficulties": 8
}

// mock category stats for different difficulties
export const mockCategoryStats: TStat = {
  "Entertainment: Film": 2,
  "Entertainment: Video Games": 2,
  "General Knowledge": 2,
  "Entertainment: Music": 2,
  "all categories": 8
}

export const mockEasyStats: TStat = {
  "Entertainment: Film": 1,
  "Entertainment: Video Games": 1,
  "General Knowledge": 1,
  "all categories": 3
}

export const mockHardStats: TStat = {
  "Entertainment: Music": 1,
  "Entertainment: Video Games": 1,
  "all categories": 2
}

export const mockMediumStats: TStat = {
  "Entertainment: Film": 1,
  "General Knowledge": 1,
  "Entertainment: Music": 1,
  "all categories": 3
}

// mock question
export const mockQuestion: IQuestion = {
  question: "Mock &lt;Question&gt;",
  id: "mockId",
  difficulty: "easy",
  correctAnswer: "Mock Correct Answer",
  category: "Mock Category",
  answers: [
    "Mocha &amp; Chai",
    "React &#124; Vue &#124; Angular",
    "Mock Aswer 3",
    "Mock Correct Answer",
  ],
};
