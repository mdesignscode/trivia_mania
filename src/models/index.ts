import sequelize from 'utils/sequelize';
import CategoryStat from './CategoryStat';
import type { TCategoryStatAttributes } from './CategoryStat';
import { type TNewStatsInput, NewStatsSchema, CategoryStatSchema } from './CategoryStatSchema';
import Question from './Question';
import type { TQuestionAttributes } from './Question';
import User from './User';
import type { TUser, TUserAttributes } from './User';
import Session from './Session';
import EmailVerification from './EmailVerification';
import UserStats from './UserStats';
import type { TUserStatsAttributes } from './UserStats';

Session.belongsTo(User);

User.hasMany(Session);
User.hasOne(UserStats);

UserStats.hasMany(CategoryStat);
UserStats.belongsTo(User);

CategoryStat.belongsTo(UserStats);

export {
        sequelize,
        Question,
        User,
        CategoryStat,
        Session,
        EmailVerification,
        UserStats,
        CategoryStatSchema,
        NewStatsSchema,
};

export type {
        TQuestionAttributes,
        TUserAttributes,
        TUser,
        TUserStatsAttributes,
        TCategoryStatAttributes,
        TNewStatsInput,
};

