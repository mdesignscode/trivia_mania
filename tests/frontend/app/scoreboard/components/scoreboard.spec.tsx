import Board from "@/app/scoreboard/components/scoreboard";
import { initialStat } from "@/models/interfaces";
import User from "@/models/user";
import { mockInitialProgress } from "@/utils/mockData";
import { mockUser } from "@/utils/test_global_context";
import { screen, render } from "@/utils/test_utils";

beforeEach(() => {
  mockUser.stats = mockInitialProgress;
});

afterAll(() => {
  mockUser.stats = initialStat;
});

// mock props
export const mockTopTenUsers = [
  mockUser,
  {
    ...mockUser,
    id: "mockUser2",
    username: "mock user 2",
    stats: mockInitialProgress,
  } as User,
];

describe("Board component", () => {
  it("Should render a list of top ten users' stats", async () => {
    // render component
    render(<Board topTenUsers={mockTopTenUsers} />);

    // get elements from DOM
    const container = await screen.findByTestId("scoreboard-container"),
      text = await screen.findByText("Top 10 Scoreboard"),
      stat1 = await screen.findByTestId("user-0-stat"),
      stat2 = await screen.findByTestId("user-1-stat");

    // assert elements are in DOM
    expect(container).toBeInTheDocument();
    expect(text).toBeInTheDocument();
    expect(stat1).toBeInTheDocument();
    expect(stat2).toBeInTheDocument();

    // get categories stats
    const scienceStats = await screen.findByTestId("mock user-Science");
    expect(scienceStats).toBeInTheDocument();
    const scienceStats2 = await screen.findByTestId("mock user 2-Science");
    expect(scienceStats2).toBeInTheDocument();

    expect(scienceStats).toHaveTextContent(/Science Correct AnswersEasy:1/)
  });

  it("snapshot matches", () => {
    // render component
    const { baseElement } = render(<Board topTenUsers={mockTopTenUsers} />);
    expect(baseElement).toMatchSnapshot();
  });
});
