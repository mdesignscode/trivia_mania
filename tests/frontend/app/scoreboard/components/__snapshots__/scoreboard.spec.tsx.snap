// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board component snapshot matches 1`] = `
<body>
  <div>
    <div
      class="p-4"
      data-testid="scoreboard-container"
    >
      <h1
        class="text-2xl font-semibold mb-4"
      >
        Top 10 Scoreboard
      </h1>
      <div
        class="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-4"
      >
        <div
          class="dark:bg-accent-200 max-h-72 overflow-y-auto p-4 rounded-lg shadow-xl dark:border-transparent border-2 col gap-3"
          data-testid="user-0-stat"
        >
          <div
            class="flex items-center"
          >
            <div
              class="mr-4"
            >
              <img
                alt="Avatar for mock user"
                class="rounded-full"
                data-nimg="1"
                decoding="async"
                height="50"
                loading="lazy"
                src="/_next/image?url=%2Favatar.png&w=128&q=75"
                srcset="/_next/image?url=%2Favatar.png&w=64&q=75 1x, /_next/image?url=%2Favatar.png&w=128&q=75 2x"
                style="color: transparent;"
                width="50"
              />
            </div>
            <div>
              <p
                class="text-xl font-semibold"
              >
                1
                . 
                mock user
              </p>
              <p
                class="text-gray-500"
              >
                Total Correct Answers: 
                1
              </p>
            </div>
          </div>
          <div
            class="flex gap-2 flex-col"
          >
            <div
              class="col"
              data-testid="mock user-Science"
            >
              <h3
                class="text-gray-700"
              >
                Science
                 Correct Answers
              </h3>
              <div
                class="flex gap-3"
              >
                <div
                  class="flex gap-1"
                  style="color: green;"
                >
                  <h4>
                    Easy
                    :
                  </h4>
                  <p>
                    1
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="dark:bg-accent-200 max-h-72 overflow-y-auto p-4 rounded-lg shadow-xl dark:border-transparent border-2 col gap-3"
          data-testid="user-1-stat"
        >
          <div
            class="flex items-center"
          >
            <div
              class="mr-4"
            >
              <img
                alt="Avatar for mock user 2"
                class="rounded-full"
                data-nimg="1"
                decoding="async"
                height="50"
                loading="lazy"
                src="/_next/image?url=%2Favatar.png&w=128&q=75"
                srcset="/_next/image?url=%2Favatar.png&w=64&q=75 1x, /_next/image?url=%2Favatar.png&w=128&q=75 2x"
                style="color: transparent;"
                width="50"
              />
            </div>
            <div>
              <p
                class="text-xl font-semibold"
              >
                2
                . 
                mock user 2
              </p>
              <p
                class="text-gray-500"
              >
                Total Correct Answers: 
                1
              </p>
            </div>
          </div>
          <div
            class="flex gap-2 flex-col"
          >
            <div
              class="col"
              data-testid="mock user 2-Science"
            >
              <h3
                class="text-gray-700"
              >
                Science
                 Correct Answers
              </h3>
              <div
                class="flex gap-3"
              >
                <div
                  class="flex gap-1"
                  style="color: green;"
                >
                  <h4>
                    Easy
                    :
                  </h4>
                  <p>
                    1
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
